packages/pigeon/README.md

```diff
@@ -1,13 +1,354 @@
+   ),
+)
+```
+
+#### Add Windows support
+
+To instruct Pigeon to generate C++ code for your app on Windows, provide paths
+for the `cppHeaderOut` and `cppSourceOut` parameters and, optionally, a
+`CppOptions` instance.
+
+```dart
+@ConfigurePigeon(
+   PigeonOptions(
+      ...
+      cppHeaderOut: 'windows/runner/messages.g.h',
+      cppSourceOut: 'windows/runner/messages.g.cpp',
+      cppOptions: CppOptions(namespace: 'pigeon_example'),
+   ),
+)
+```
+
+#### Add Linux support
+
+To instruct Pigeon to generate GObject code for your app on Linux, provide paths
+for the `gobjectHeaderOut` and `gobjectSourceOut` parameters and, optionally, a
+`GObjectOptions` instance.
+
+```dart
+@ConfigurePigeon(
+   PigeonOptions(
+      ...
+      gobjectHeaderOut: 'linux/messages.g.h',
+      gobjectSourceOut: 'linux/messages.g.cc',
+      gobjectOptions: GObjectOptions(),
+   ),
+)
+```
+
+#### A note on Web support
+
+Pigeon does not support the Web because Flutter apps compiled to the Web can
+already directly call any JavaScript code without switching threads, as is
+currently required in Flutter on mobile or desktop. Pigeon generates code which
+performs two roles:
+
+1. Uses Flutter's `MethodChannel`s concept to jump from the UI thread to the
+Platform thread (or in the other direction), and
+2. Marshalls data between Dart and the native language.
+
+Flutter Web apps intrinsically do not encounter the first problem and only
+encounter a form of the second problem if you need to interface with a
+library's `d.ts` interface. In this case, you may need to author custom Dart
+classes which match those TypeScript definitions. Check pub.dev for existing solutions.
+
+### Run the builder
+
+Once your Pigeon files define any data classes and functions you wish to invoke,
+you can run the builder:
+
+```sh
+$ dart run pigeon --input pigeons/messages.dart
+```
+
+You must run this command once per Pigeon file, specifying a single `--input` each
+time. Pigeon files may only import `package:pigeon/pigeon.dart`, so they cannot
+depend on each other. Pigeon files must also all have their own
+`@ConfigurePigeon(PigeonOptions())` declaration.
+
+### Use the generated code
+
+You should now see matching output files at the locations you specified in your
+`PigeonOptions` instance, or in your command line arguments. The two primary
+scenarios to explore are calling native code from Dart and calling Dart code from
+native.
+
+#### Calling native code from Dart
+
+The sample code in this Quickstart defines an example native API named `ExampleHostApi`.
+Pigeon will generate a complete Dart implementation and the equivalent of an abstract
+class in each native language (for example, a `protocol` in Swift and an `interface`
+in Kotlin).
+
+See the [language-specific guides](example/README.md#HostApi-Example) for help instantiating the native classes
+generated by Pigeon.
+
+In Dart, typical use within a `StatefulWidget` might look like this:
```

Comment:
I don't think we have any code in our repo that looks like this, so I would not say this is typical use. I'm not sure why we need an example for this at all.

Reply:
[This code](https://github.com/flutter/packages/blob/main/packages/pigeon/example/app/lib/main.dart#L55-L59) (and on down) is what inspired it.

FWIW, this whole PR was inspired by me reading Pigeon's documentation, still being quite confused about the high level picture, and then having to piece it together from sample source code. My goal is to elevate the high level picture with snippets to the top-level README.

Reply:
> I'm not sure why we need an example for this at all.

Pigeon is great, but I thinks fundamental API shape has a decent learning curve (writing a pseudo-class, generating an abstract class off that, defining a concrete class and registering it with the generated abstract class). It's all fairly unusual, so my general sense here is that we're currently short of context and perspective explainers around code samples.

Reply:
> [This code](https://github.com/flutter/packages/blob/main/packages/pigeon/example/app/lib/main.dart#L55-L59) (and on down) is what inspired it.

That code is pretty artificial, due to the constraints of trying to fit an example of Pigeon usage into a single-file application.

> thinks fundamental API shape has a decent learning curve (writing a pseudo-class, generating an abstract class off that, defining a concrete class and registering it with the generated abstract class). It's all fairly unusual, so my general sense here is that we're currently short on context and perspective explainers around code samples.

Sure, but none of that is related to `StatefulWidget`. I would rather the readme focus on important concepts that are fundamental to using Pigeon, rather than spending space on the unrealistic example of embedding usage directly into a `StatefulWidget`.